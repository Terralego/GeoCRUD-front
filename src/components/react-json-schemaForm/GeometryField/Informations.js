import React, { useMemo, memo } from 'react';
import { Trans, useTranslation } from 'react-i18next';
import PropTypes from 'prop-types';
import Message from '../../Message';
import { useGeometryField } from './GeometryFieldProvider';

const getIconFromType = type => {
  if (type.includes('Point')) {
    return 'point';
  }
  if (type.includes('Line')) {
    return 'line';
  }
  return 'polygon';
};

const Information = ({
  schema: {
    default: { coordinates: schemaCoordinates, type: schemaType },
  },
}) => {
  const { isRequired, isRouting, nextFormData: { geom } = {} } = useGeometryField();

  const { t } = useTranslation();

  const { coordinates = schemaCoordinates, type = schemaType } = geom || {};

  const action = !coordinates.length ? 'create' : 'edit';

  const helperText = useMemo(() => {
    if (isRouting) {
      return t('jsonSchema.geometryField.helper-routing');
    }
    if (action === 'create') {
      return (
        <Trans i18nKey="jsonSchema.geometryField.helper">
          By using the tool
          <span
            className={`mapbox-gl-draw_ctrl-draw-btn mapbox-gl-draw_${getIconFromType(type)}`}
          />
          on map
        </Trans>
      );
    }
    return t('jsonSchema.geometryField.helper-edit', { type });
  }, [action, isRouting, t, type]);

  return (
    <>
      <p className="control-label">
        <span>
          {t(`jsonSchema.geometryField.information-${action}`, { type })}
          {isRequired && <span className="required">*</span>}
        </span>
      </p>
      <Message className="geometry-field__message" intent="primary">
        {helperText}
      </Message>
    </>
  );
};

Information.propTypes = {
  schema: PropTypes.shape({
    default: PropTypes.shape({
      coordinates: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.number, PropTypes.array])),
      type: PropTypes.string,
    }),
  }),
};

Information.defaultProps = {
  schema: {
    default: {
      coordinates: [],
      type: undefined,
    },
  },
};

export default memo(Information);
