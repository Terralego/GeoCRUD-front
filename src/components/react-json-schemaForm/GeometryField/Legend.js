import React, { memo } from 'react';
import { useTranslation } from 'react-i18next';
import { useGeometryField } from './GeometryFieldProvider';

const Legend = ({ title, ...props }) => {
  const { t } = useTranslation();
  const { isRequired, isRequiredInEditView } = useGeometryField();

  const mandatoryField = isRequired ? ' *' : null;
  const mandatoryLegend = isRequiredInEditView ? (
    <span className="details__list-edit-mandatory">{t('CRUD.details.mandatory')}</span>
  ) : null;

  return (
    <legend {...props}>
      <span>
        {title}
        {mandatoryField}
      </span>
      {mandatoryLegend}
    </legend>
  );
};

export default memo(Legend);
