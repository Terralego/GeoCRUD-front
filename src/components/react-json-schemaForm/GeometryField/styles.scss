@import '~@blueprintjs/core/src/components/forms/common';

.CRUD-edit {
  .geometry-field {
    position: relative;
    text-align: center;
    &__or {
      display: flex;
      width: 100%;
      align-items: center;
      margin: 1rem 0;
      span {
        padding: 0 0.5rem;
      }
      &::before,
      &::after {
        content: '';
        flex-grow: 1;
        margin-top: 2px;
        border-top: 1px solid #ddd;
      }
    }
    &__col {
      border: solid #ddd;
      border-width: 1px 0;
      margin: 0.5rem 0;
      padding: 1rem 0;
    }

    &__row {
      padding: 0.5rem;
    }
    &__message {
      max-width: max-content;
      margin: auto;
    }
    .control-label {
      font-size: 1.1rem;
      display: flex;
      align-items: center;
      justify-content: center;
      > * {
        margin: 0 0.2rem;
      }
    }
    &__reset {
      position: absolute;
      bottom: 0.5rem;
      right: 0;
    }
  }
  .mapbox-gl-draw_ctrl-draw-btn {
    display: inline-block;
    background-color: #fff;
    margin: 0 2px;
    width: 29px;
    height: 29px;
    vertical-align: middle;
    border-radius: 4px;
    box-shadow: 0 0 0 2px rgba(0, 0, 0, 0.1);
  }
}

.details__list-edit {
  .geometry-field {
    &__col {
      margin-top: 1.8rem;
    }
  }
}

.mapbox-gl-path-container {
  float: left;
  pointer-events: auto;
  user-select: none;
  background-color: #ffffff;
  color: #182026;
  padding: 4px 8px;
  margin: 10px;
  border-radius: 4px;
  box-shadow: 0 0 0 2px rgba(0, 0, 0, 0.1);
}

.mapbox-gl-path-theme-selection__checkbox-label {
  cursor: pointer;
  padding-left: 5px;
}

.mapbox-gl-path-theme-selection {
  height: 0;
  opacity: 0;
  overflow: hidden;
  transition: all 0.3ms ease;
}

.mapboxgl-popup-content {
  padding: 1.2rem 0.8rem 0.8rem;
}

.mapboxgl-popup-close-button {
  padding: 0 0.5rem;
}

.mapbox-gl-path-popup-button {
  appearance: none;
  cursor: pointer;
  display: block;
  border: none;
  margin: 0.3rem 0;
  background: transparent;
  font-size: 0.8rem;
  text-align: left;
}

.mapbox-gl-path-popup-button:hover,
.mapbox-gl-path-popup-button:focus {
  color: #000;
  text-decoration: underline;
  border: 0;
}

.mapbox-gl-path-popup-button::before {
  float: left;
  margin-top: -0.1rem;
  margin-right: 0.2rem;
  vertical-align: middle;
}

.mapbox-gl-path-popup-delete::before {
  content: url('./icons/delete-point.svg');
}

.mapbox-gl-path-popup-createPointOnLine::before {
  content: url('./icons/create-point.svg');
}

.mapbox-gl-path-popup-createIntermediatePointOnLine::before {
  content: url('./icons/create-intermediate-point.svg');
}

.mapbox-gl-path-popup-loop::before {
  content: url('./icons/loop.svg');
}

.mapbox-gl-path-popup-oneWay::before {
  content: url('./icons/un-loop.svg');
}

.mapbox-gl-path-popup-disableFollowDirectionMode::before {
  content: url('./icons/un-follow-direction.svg');
}

.mapbox-gl-path-popup-enableFollowDirectionMode::before {
  content: url('./icons/follow-direction.svg');
}

.mapbox-gl-path-theme-selection__checkbox:checked ~ .mapbox-gl-path-theme-selection {
  height: auto;
  padding: 0.3rem 0;
  opacity: 1;
  margin: 0.6rem 0 0;
  border-top: 1px solid #eaeaea;
}

.mapbox-gl-path-theme-selection__radio-label {
  margin: 0.2rem 0.5rem;
  display: inline-block;
  cursor: pointer;
}
