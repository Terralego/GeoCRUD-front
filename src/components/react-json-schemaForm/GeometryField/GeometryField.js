import React from 'react';
import { useTranslation } from 'react-i18next';

import PropTypes from 'prop-types';

import GeometryFieldProvider from './GeometryFieldProvider';

import ImportGeomFile from '../../ImportGeomFile';

import MapInteraction from './MapInteraction';
import Informations from './Informations';
import Legend from './Legend';
import ResetGeometry from './ResetGeometry';
import './styles.scss';

const GeometryField = ({ formData, name, onChange, schema }) => {
  const { t } = useTranslation();

  return (
    <GeometryFieldProvider formData={formData} onChange={onChange} name={name} schema={schema}>
      <fieldset className="geometry-field">
        <Legend className="geometry-field__legend" title={schema.title} />
        <div className="geometry-field__col">
          <MapInteraction />
          <div className="form-group field">
            <div className="geometry-field__row">
              <Informations schema={schema} />
            </div>
            <span className="geometry-field__or">
              <span>{t('jsonSchema.geometryField.or')}</span>
            </span>
            <div className="geometry-field__row">
              <ImportGeomFile />
            </div>
          </div>
        </div>
        <ResetGeometry intent="danger" schema={schema} />
      </fieldset>
    </GeometryFieldProvider>
  );
};

GeometryField.propTypes = {
  formData: PropTypes.shape({
    coordinates: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.number, PropTypes.array])),
    type: PropTypes.string,
  }),
  name: PropTypes.string,
  onChange: PropTypes.func,
  schema: PropTypes.shape({}),
};

GeometryField.defaultProps = {
  formData: {
    coordinates: [],
    type: '',
  },
  name: undefined,
  onChange: () => {},
  schema: undefined,
};

export default GeometryField;
