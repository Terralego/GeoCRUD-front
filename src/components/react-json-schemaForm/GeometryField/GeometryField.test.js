import React from 'react';
import renderer from 'react-test-renderer';

import GeometryField from './GeometryField';

jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: text => text,
  }),
}));

jest.mock('./GeometryFieldProvider', () => props => <div {...props} />);
jest.mock('../../ImportGeomFile', () => props => <div {...props} />);
jest.mock('./MapInteraction', () => props => <div {...props} />);
jest.mock('./ResetGeometry', () => props => <div {...props} />);
jest.mock('./Informations', () => props => <div {...props} />);
jest.mock('./Legend', () => props => <div {...props} />);

it('should snapshot correctly', () => {
  const tree = renderer.create(<GeometryField schema={{ title: 'foo' }} />).toJSON();
  expect(tree).toMatchSnapshot();
});
