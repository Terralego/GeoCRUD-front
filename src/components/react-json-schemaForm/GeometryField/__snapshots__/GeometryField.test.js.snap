// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should snapshot correctly 1`] = `
<div
  formData={
    Object {
      "coordinates": Array [],
      "type": "",
    }
  }
  onChange={[Function]}
  schema={
    Object {
      "title": "foo",
    }
  }
>
  <fieldset
    className="geometry-field"
  >
    <div
      className="geometry-field__legend"
      title="foo"
    />
    <div
      className="geometry-field__col"
    >
      <div />
      <div
        className="form-group field"
      >
        <div
          className="geometry-field__row"
        >
          <div
            schema={
              Object {
                "title": "foo",
              }
            }
          />
        </div>
        <span
          className="geometry-field__or"
        >
          <span>
            jsonSchema.geometryField.or
          </span>
        </span>
        <div
          className="geometry-field__row"
        >
          <div />
        </div>
      </div>
    </div>
    <div
      intent="danger"
      schema={
        Object {
          "title": "foo",
        }
      }
    />
  </fieldset>
</div>
`;
